ffconv/stream_processors.py

```diff
@@ -0,0 +1,240 @@
+
+    def process(self):
+        """
+        Process this stream, which might mean converting it or not.
+
+        :return: processed stream data
+        """
+        log('Processing stream <index:{} type:{} codec:{}>...'.format(self.index, self.media_type, self.codec), 2)
+        if self.must_convert:
+            # Must convert, run conversion and clean-up
+            log('Converting...', 3)
+            self.convert()
+            log('Cleaning up file...', 3)
+            self.clean_up()
+            self.input = self.output
+            self.index = 0
+            log('Done', 2)
+
+        else:
+            # Nothing to do
+            log('Skipping, no conversion required', 3)
+
+        # Build and return data: input file, index and language (used by merger)
+        res = {'input': self.input, 'index': self.index}
+        if self.language:
+            res['language'] = self.language
+        return res
+
+
+class VideoProcessor(StreamProcessor):
+    """
+    Video stream processor, provides specific video conversion functionality.
+    """
+    media_type = 'video'
+
+    def _init_stream(self, input, stream, profile):
+        """
+        Set video-specific input and target specs,
+        """
+        # Set input reference frames value
+        self.refs = int(stream['refs'])
+
+        # Set target values for frames, profile, level, preset and quality
+        height = int(stream.get('height', 720))
+        self.max_refs = profile[self.media_type]['max_refs'].get(height, 5)
+        self.target_profile = profile[self.media_type]['profile']
+        self.target_level = profile[self.media_type]['level']
+        self.target_preset = profile[self.media_type]['preset']
+        self.target_quality = profile[self.media_type]['quality']
+
+    @property
+    def must_convert(self):
+        """
+        Conversion check: besides base check (codec compatibility), check that
+        reference frames are acceptable.
+        """
+        return any((super(VideoProcessor, self).must_convert,
+                    self.refs > self.max_refs))
+
+    def clean_up(self):
+        """
+        Cleanup is no-op for video, because the stream is not modified.
+        """
+        pass
+
+    def convert(self):
+        """
+        Convert the video stream according to the selected target values,
+        creating a temporary video file to use by merger.
+        """
+        cmd = ['ffmpeg', '-i', self.input, '-map', '0:{}'.format(self.index),
+               '-c:v', self.target_codec, '-preset', str(self.target_preset),
+               '-crf', str(self.target_quality), '-profile:v',
+               self.target_profile, '-level', self.target_level, self.output]
+        execute_cmd(cmd)
+
+
+class AudioProcessor(StreamProcessor):
+    """
+    Audio stream processor, provides specific audio conversion functionality.
+    """
+    media_type = 'audio'
+
+    def _init_stream(self, input, stream, profile):
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Redefining built-in 'input' [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://localhost:9000/coding_rules#rule_key=Pylint%3AW0622)